src/main/java/org/apache/maven/plugins/deploy/DeployFileMojo.java

```diff
@@ -274,9 +274,7 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                 classifier = artifactType.getClassifier();
             }
         }
-        Artifact mainArtifact = new DefaultArtifact(
-                        groupId, artifactId, classifier, isFilePom ? "pom" : getExtension(file), version)
-                .setFile(file);
+        Artifact mainArtifact = new DefaultArtifact(groupId, artifactId, classifier, packaging, version).setFile(file);
```

Comment:
Please look at java doc: https://maven.apache.org/resolver/apidocs/org/eclipse/aether/artifact/DefaultArtifact.html#%3Cinit%3E(java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String)

fourth parameter is a file extension not packaging

Reply:
instead of `getExtension(file)` we can use `artifactType.getExtension()` as improvements

Reply:
> Please look at java doc: https://maven.apache.org/resolver/apidocs/org/eclipse/aether/artifact/DefaultArtifact.html#%3Cinit%3E(java.lang.String,java.lang.String,java.lang.String,java.lang.String,java.lang.String)
> 
> fourth parameter is a file extension not packaging

But for some reason, before 16541da43c51237cf7183c4500a576108ad946b8, packaging _was_ used as the file extension. Are you saying that was never intended and was a bug?

If so, is there any way to get the old behavior back in maven-deploy-plugin 3.0?

`artifactType.getExtension()` won't solve the problem, as this is only working known `packaging`s. It does not cover the case `-Dpackaging=completelarbitrary` where no `artifactType` is known.

I have uploaded an alternative proposal: If we have an `artifactType` use its extension, otherwise fall back to `packaging` for unknown types.

Reply:
According to the [documentation|https://maven.apache.org/plugins/maven-deploy-plugin/deploy-file-mojo.html], `packaging` designate the “Type of the artifact to be deployed. Retrieved from the <packaging> element of the POM file if a POM file specified. Defaults to the file extension if it is not specified via command line or POM.”

The way I read it, it means that the `packaging` tag takes precedence (whether from command line or from the POM), followed by the extension of the file to upload as a fallback.

I just hit this issue (MDEPLOY-311) because we generate a zip file using the truezip plugin (that relies on the extension) and then we need to deploy using a different extension.
The `.epub` ebooks (used for documentation) are really `.zip` files with a specific content but this is true for quite a lot of other formats as well so there really is a use-case for using `<packaging>` with a different extension than the one on the deployed file.